evap/staff/importers/json.py

```diff
@@ -0,0 +1,448 @@
+        return log
+
+    def send_mail(self):
+        subject = "[EvaP] JSON importer log"
+
+        managers = UserProfile.objects.filter(groups__name="Manager", email__isnull=False)
+        if not managers:
+            return
+        mail = EmailMultiAlternatives(
+            subject,
+            self.get_log(),
+            settings.SERVER_EMAIL,
+            [manager.email for manager in managers],
+        )
+        mail.send()
+
+
+class JSONImporter:
+    DATETIME_FORMAT = "%d.%m.%Y %H:%M:%S"
+
+    def __init__(self, semester: Semester, default_course_end: str) -> None:
+        self.semester = semester
+        self.default_course_end = default_course_end
+        self.user_profile_map: dict[str, UserProfile] = {}
+        self.course_type_cache: dict[str, CourseType] = {
+            import_name.strip().lower(): course_type
+            for course_type in CourseType.objects.all()
+            for import_name in course_type.import_names
+        }
+        self.program_cache: dict[str, Program] = {
+            import_name.strip().lower(): program
+            for program in Program.objects.all()
+            for import_name in program.import_names
+        }
+        self.course_map: dict[str, Course] = {}
+        self.statistics = ImportStatistics()
+
+    def _get_users_with_longest_title(self, user_profiles: list[UserProfile]) -> list[UserProfile]:
+        user_profiles_by_len = unordered_groupby((len(user.title), user) for user in user_profiles)
+        for key in sorted(user_profiles_by_len.keys(), reverse=True):
+            return user_profiles_by_len[key]
+        return user_profiles_by_len[max(user_profiles_by_len.keys())] if user_profiles_by_len else []
+
+    def _remove_non_responsible_users(self, user_profiles: list[UserProfile]) -> list[UserProfile]:
+        return list(filter(lambda p: p.email not in settings.NON_RESPONSIBLE_USERS, user_profiles))
+
+    def _get_course_type(self, name: str) -> CourseType:
+        lookup = name.strip().lower()
+        if lookup in self.course_type_cache:
+            return self.course_type_cache[lookup]
+
+        # It could happen that the importer needs a new course type
+        course_type, __ = CourseType.objects.get_or_create(name_de=name, defaults={"name_en": name})
+        self.course_type_cache[name] = course_type
+        return course_type
+
+    def _get_program(self, name: str) -> Program:
+        lookup = name.strip().lower()
+        if lookup in self.program_cache:
+            return self.program_cache[lookup]
+
+        # It could happen that the importer needs a new program
+        program, __ = Program.objects.get_or_create(name_de=name, defaults={"name_en": name})
+        self.program_cache[name] = program
+        return program
+
+    def _get_user_profiles(self, data: list[ImportRelated]) -> list[UserProfile]:
+        return [
+            self.user_profile_map[related["gguid"]] for related in data if related["gguid"] in self.user_profile_map
```

Comment:
Can it happen that such an id doesn't exist? And why do we ignore it if that happens?

Reply:
Can you add this explanation as a comment?